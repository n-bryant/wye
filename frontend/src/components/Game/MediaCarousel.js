import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";

import Icon from "@mdi/react";
import { mdiPlay } from "@mdi/js";

import styles from "./MediaCarousel.styles";

export const MEDIA_TYPES = {
  VIDEO: "video",
  IMG: "img"
};

/**
 * renders a media carousel with a featured item
 */
export const MediaCarousel = props => {
  const classnames = MediaCarousel.classnames(props);
  const { media } = props;

  const initialState = {
    item: media[0],
    type: media[0].title ? MEDIA_TYPES.VIDEO : MEDIA_TYPES.IMG
  };
  const [featuredItem, setFeaturedItem] = React.useState(initialState);

  // build an img or video based on the featuredItem's type
  const Feature = () => {
    const { item } = featuredItem;
    if (featuredItem.type === MEDIA_TYPES.VIDEO) {
      return (
        <video
          className={classnames.element("featuredItem")}
          playsInline={true}
          src={item.fullsizeUrl}
          poster={item.thumbnail}
          autoPlay
          preload={"none"}
          controls
          muted
        >
          {item.title}
        </video>
      );
    } else {
      const [loaded, setLoaded] = React.useState(false);

      const onLoad = () => {
        setLoaded(true);
      };
      return (
        <img
          className={classnames.element("featuredItem", {
            hidden: !loaded
          })}
          src={item.fullsizeUrl}
          alt="featured item"
          onLoad={onLoad}
        />
      );
    }
  };

  return (
    <Grid
      className={classnames.root()}
      container
      item
      xs={12}
      md={7}
      spacing={1}
    >
      <Grid
        className={classnames.element("featuredItemContainer")}
        item
        xs={12}
      >
        <div className={classnames.element("featuredItemWrapper")}>
          <img
            className={classnames.element("featuredItemSpacer")}
            src="/featuredItemSpacer.png"
          />
          <Feature />
        </div>
      </Grid>
      <Grid
        className={classnames.element("thumbnailsContainer")}
        container
        wrap="nowrap"
        item
        xs={12}
        spacing={1}
      >
        {media.map(item => (
          <Grid
            key={item.id}
            item
            className={classnames.element("setFeaturedButtonContainer")}
          >
            <Button
              className={classnames.element("setFeaturedButton")}
              onClick={() =>
                setFeaturedItem({
                  item,
                  type: item.title ? MEDIA_TYPES.VIDEO : MEDIA_TYPES.IMG
                })
              }
            >
              <img
                className={classnames.element("thumbnail")}
                src={item.thumbnailUrl}
                alt={item.id}
              />
              {item.title && (
                <div className={classnames.element("playIconContainer")}>
                  <Icon
                    className={classnames.element("playIcon")}
                    path={mdiPlay}
                  />
                </div>
              )}
            </Button>
          </Grid>
        ))}
      </Grid>
    </Grid>
  );
};
MediaCarousel.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}MediaCarousel`
);
MediaCarousel.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    featuredItemContainer: PropTypes.string,
    featuredItemSpacer: PropTypes.string,
    featuredItem: PropTypes.string,
    featuredItemHidden: PropTypes.string,
    featuredItemWrapper: PropTypes.string,
    thumbnailsContainer: PropTypes.string,
    setFeaturedButtonContainer: PropTypes.string,
    setFeaturedButton: PropTypes.string,
    thumbnail: PropTypes.string,
    playIconContainer: PropTypes.string,
    playIcon: PropTypes.string
  }),
  // details for the media to display
  media: PropTypes.array
};
MediaCarousel.defaultProps = {
  classes: {}
};

export default withStyles(styles)(MediaCarousel);
