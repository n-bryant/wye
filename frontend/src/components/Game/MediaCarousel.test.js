import Icon from "@mdi/react";
import { mdiPlay } from "@mdi/js";
import Button from "@material-ui/core/Button";
import { MediaCarousel, MEDIA_TYPES } from "./MediaCarousel";

describe("MediaCarousel", () => {
  const props = {
    classes: {
      root: "root",
      featuredItemContainer: "featuredItemContainer",
      featuredItemSpacer: "featuredItemSpacer",
      featuredItem: "featuredItem",
      featuredItemHidden: "featuredItemHidden",
      featuredItemWrapper: "featuredItemWrapper",
      thumbnailsContainer: "thumbnailsContainer",
      setFeaturedButtonContainer: "setFeaturedButtonContainer",
      setFeaturedButton: "setFeaturedButton",
      thumbnail: "thumbnail",
      playIconContainer: "playIconContainer",
      playIcon: "playIcon"
    },
    media: [
      {
        id: "1",
        thumbnailUrl: "thumb1",
        fullsizeUrl: "full1",
        title: "title1"
      },
      {
        id: "2",
        thumbnailUrl: "thumb2",
        fullsizeUrl: "full2"
      }
    ]
  };

  let setFeaturedItem;
  beforeEach(() => {
    const featuredItem = {
      item: {
        ...props.media[0]
      },
      type: MEDIA_TYPES.VIDEO
    };
    setFeaturedItem = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [featuredItem, setFeaturedItem]);
  });

  it("should render successfully", () => {
    const wrapper = shallow(<MediaCarousel {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should display a play icon on video thumbnails", () => {
    const wrapper = shallow(<MediaCarousel {...props} />);
    expect(
      wrapper
        .find(Button)
        .at(0)
        .find(Icon)
        .prop("path")
    ).toBe(mdiPlay);
    expect(
      wrapper
        .find(Button)
        .at(1)
        .find(Icon).length
    ).toBe(0);
  });

  it("should update the featured item when a thumbnail is clicked", () => {
    const wrapper = shallow(<MediaCarousel {...props} />);
    wrapper
      .find(Button)
      .at(1)
      .prop("onClick")();
    expect(setFeaturedItem).toHaveBeenCalledWith({
      item: props.media[1],
      type: MEDIA_TYPES.IMG
    });
  });
});
