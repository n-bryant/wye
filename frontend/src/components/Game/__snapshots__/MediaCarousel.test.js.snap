// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaCarousel should render successfully 1`] = `
<WithStyles(ForwardRef(Grid))
  className="wye-MediaCarousel root"
  container={true}
  item={true}
  md={7}
  spacing={1}
  xs={12}
>
  <WithStyles(ForwardRef(Grid))
    className="wye-MediaCarousel__featuredItemContainer featuredItemContainer"
    item={true}
    xs={12}
  >
    <div
      className="wye-MediaCarousel__featuredItemWrapper featuredItemWrapper"
    >
      <img
        className="wye-MediaCarousel__featuredItemSpacer featuredItemSpacer"
        src="/featuredItemSpacer.png"
      />
      <Feature />
    </div>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    className="wye-MediaCarousel__thumbnailsContainer thumbnailsContainer"
    container={true}
    item={true}
    spacing={1}
    wrap="nowrap"
    xs={12}
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-MediaCarousel__setFeaturedButtonContainer setFeaturedButtonContainer"
      item={true}
      key="1"
    >
      <WithStyles(ForwardRef(Button))
        className="wye-MediaCarousel__setFeaturedButton setFeaturedButton"
        onClick={[Function]}
      >
        <img
          alt="1"
          className="wye-MediaCarousel__thumbnail thumbnail"
          src="thumb1"
        />
        <div
          className="wye-MediaCarousel__playIconContainer playIconContainer"
        >
          <Icon
            className="wye-MediaCarousel__playIcon playIcon"
            color="currentColor"
            horizontal={false}
            path="M8,5.14V19.14L19,12.14L8,5.14Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </div>
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="wye-MediaCarousel__setFeaturedButtonContainer setFeaturedButtonContainer"
      item={true}
      key="2"
    >
      <WithStyles(ForwardRef(Button))
        className="wye-MediaCarousel__setFeaturedButton setFeaturedButton"
        onClick={[Function]}
      >
        <img
          alt="2"
          className="wye-MediaCarousel__thumbnail thumbnail"
          src="thumb2"
        />
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
